_posts/2017-11-22-macros.md

```diff
@@ -0,0 +1,305 @@
+as well as projects making heavy use of macros for embedded DSLs.
+
+If there is interest, I'm happy to write a guide on how to run the analysis on
+closed-source projects.
+It's quite easy to do, and the results are anonymous excluding the signature
+names of the invoked def macros, which can be manually obfuscated.
+A large and representative corpus of "real-world" Scala codebases will help us
+continue to use a data-driven approach to prioritize the development of new
+macros.
+
+## Tree transformations are tricky
+
+Each of the diverse macro applications listed above require different sets of
+features from the macro system.
+For example, code generation macros typically only require inspection on existing types
+and creation of new terms.
+Embedded DSLs on the other hand typically require inspection and transformation of
+existing terms created by the compiler.
+Optimization macros like those in log4s may even be avoided with improvements
+to alternative language features like `inline`.
+
+Some features of the macro system appear to be relatively simple to support
+across different compilers.
+In the [scalacenter/macros] repository, we have a prototype macro system
+that runs on both Scala 2.x and Dotty.
+In only a few days, we got some interesting macros working for both compilers:
+a JSON automatic serializer for case classes and most of the [sourcecode] macros.
+That pretty exciting news for code generation macros and a subset of developer ergonomics macros.
+Inspection on types and creation of new terms doesn't appear to introduce deep
+technical road-blockers!
+However, we struggled to implement some other macros like [utest] assert since
+it requires the ability to transform trees.
+
+Macros that transform trees produced by the compiler are demanding on the macro
+system.
+When a term is transformed, the macro library or the macro system must ensure
+that internal invariants imposed by the compiler are respected.
+For example, the Scala compiler assumes that typed tree nodes cannot have
+untyped children.
+In scala-reflect, the burden is on the macro author to ensure this invariant
+holds.
+Mistakes from mixing untyped and typed tree nodes result in cryptic compiler
+crashes.
+In the new macro system, we want to avoid exposing such bad experience to macro
+authors and library users.
+
+This observation that tree transformation are difficult to support doesn't
+imply that new macros won't support them.
+It simply means that we need give this problem more time and thought so that we
+can come up with a robust solution.
+So far, we have seen two promising ideas to address or circumvent this problem,
+one by Martin Odersky and one by Lionel Parreaux.
+
+## Unification of LMS and Macros
+
+Martin Odersky recently shared [a gist](https://gist.github.com/odersky/f91362f6d9c58cc1db53f3f443311140)
+on Principled Metaprogramming for Scala.
+The design proposed in this gist is quite close to Meta OCaml as well as to a
```

Comment:
"MetaOCaml" is written without a space 😉

Reply:
Fixed.